src/rpc/util.cpp

```diff
@@ -634,6 +636,21 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     return ret;
 }
 
+static const UniValue& ArgOrDefault(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
+{
+    CHECK_NONFATAL(i < params.size());
+    const UniValue& arg{CHECK_NONFATAL(req)->params[i]};
+    const RPCArg::Fallback& fallback{params.at(i).m_fallback};
+    CHECK_NONFATAL(std::holds_alternative<RPCArg::Default>(fallback));
+    return arg.isNull() ? std::get<RPCArg::Default>(fallback) : arg;
+}
+
+template <>
+int RPCHelpMan::Arg<int>(size_t i) const
```

Comment:
This implementation only supports being called on parameters with a `RPCArg::Default` fallback. Do you envision this being extended to also support `RPCArg::DefaultHint` and `RPCArg::Optional::OMITTED` fallbacks, and if so, how?

I think one way is with a `std::optional<int> RPCHelpMan::Arg<std::optional<int>>(size_t i) const` specialization. I tinkered around with it a bit and this is what I came up with:

<details>
<summary>git diff on fa6b2da57ef7ff125a493c7835cb15935255ff8c</summary>

```diff
diff --git a/src/rpc/server.cpp b/src/rpc/server.cpp
index daf751111f..fd1d6c6a63 100644
--- a/src/rpc/server.cpp
+++ b/src/rpc/server.cpp
@@ -180,8 +180,8 @@ static RPCHelpMan stop()
     // Event loop will exit after current HTTP requests have been handled, so
     // this reply will get back to the client.
     StartShutdown();
-    if (jsonRequest.params[0].isNum()) {
-        UninterruptibleSleep(std::chrono::milliseconds{jsonRequest.params[0].getInt<int>()});
+    if (const auto wait{self.Arg<std::optional<int>>(0)}) {
+        UninterruptibleSleep(std::chrono::milliseconds{*wait});
     }
     return RESULT;
 },
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 4e0bace96f..c1e626eeaa 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -636,19 +636,37 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     return ret;
 }
 
-static const UniValue& ArgOrDefault(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
+static const std::optional<UniValue> ArgOrDefault(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
 {
     CHECK_NONFATAL(i < params.size());
     const UniValue& arg{CHECK_NONFATAL(req)->params[i]};
     const RPCArg::Fallback& fallback{params.at(i).m_fallback};
-    CHECK_NONFATAL(std::holds_alternative<RPCArg::Default>(fallback));
-    return arg.isNull() ? std::get<RPCArg::Default>(fallback) : arg;
+    const bool has_default{std::holds_alternative<RPCArg::Default>(fallback)};
+
+    if (arg.isNull()) {
+        if (has_default) {
+            return std::get<RPCArg::Default>(fallback);
+        } else { 
+            return std::nullopt;
+        }
+    }
+    return arg;
 }
 
 template <>
 int RPCHelpMan::Arg<int>(size_t i) const
 {
-    return ArgOrDefault(m_args, m_req, i).getInt<int>();
+    return CHECK_NONFATAL(ArgOrDefault(m_args, m_req, i)).value().getInt<int>();
+}
+
+template <>
+std::optional<int> RPCHelpMan::Arg<std::optional<int>>(size_t i) const
+{
+    if (const auto result{ArgOrDefault(m_args, m_req, i)}) {
+        return result.value().getInt<int>();
+    } else {
+        return std::nullopt;
+    }
 }
 
 bool RPCHelpMan::IsValidNumArgs(size_t num_args) const

```
</details>

I like the limited scope of this PR to implement just `int` types, but perhaps it would be useful to extend it a bit and implement the helper for non-default parameters too? At the very least we get a sense of what we want the interface to look like?

If we don't extend it, perhaps the documentation for `template <typename R> R Arg(size_t i) const;` in `rpc/util.h` needs to be updated to reflect where it can safely be used?

Reply:
Big fan of this rewrite, makes it pretty extensible without too much code duplication. Nice!

Building on your new approach, the non-default parameters could potentially be handled with pointers, to avoid unnecessary copy operations. For example, I think these 2 lines: https://github.com/bitcoin/bitcoin/blob/b565485c24c0feacae559a7f6f7b83d7516ca58d/src/wallet/rpc/spend.cpp#L274-L275 

can be rewritten as: 
```cpp
    if (auto comment{self.Arg<const std::string*>(2)}; comment && !(comment->empty())) {
        mapValue["comment"] = *comment;
    }
```

Not necessarily shorter, but much more idiomatic imo? I've implemented it like this - can probably be cleaned up quite a bit more:

<details>
<summary>git diff on fa5468dcf0109d11ddaeb4b3591e6e04b4ea6125</summary>

```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 0db66ec1b1..1f287fb291 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -636,27 +636,47 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     return ret;
 }
 
-static const UniValue& ArgOrDefault(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
+static const UniValue* ArgOrDefault(const std::vector<RPCArg>& params, const JSONRPCRequest* req, size_t i)
 {
     CHECK_NONFATAL(i < params.size());
     const UniValue& arg{CHECK_NONFATAL(req)->params[i]};
-    if (!arg.isNull()) return arg;
+    if (!arg.isNull()) return &arg;
     const RPCArg::Fallback& fallback{params.at(i).m_fallback};
-    CHECK_NONFATAL(std::holds_alternative<RPCArg::Default>(fallback));
-    return std::get<RPCArg::Default>(fallback);
+    if (std::holds_alternative<RPCArg::Default>(fallback)) {
+        return &(std::get<RPCArg::Default>(fallback));
+    }
+    return nullptr;
 }
 
-#define TMPL_INST(ret_type, get_arg, get_type)            \
-    template <>                                           \
-    ret_type RPCHelpMan::get_arg(size_t i) const          \
-    {                                                     \
-        return ArgOrDefault(m_args, m_req, i).get_type(); \
-    }                                                     \
+#define TMPL_INST(ret_type, get_arg, get_type)        \
+    template <>                                       \
+    ret_type RPCHelpMan::get_arg(size_t i) const      \
+    {                                                 \
+        return CHECK_NONFATAL(ArgOrDefault(m_args, m_req, i))->get_type(); \
+    }                                                 \
+    void force_semicolon()
+
+#define TMPL_INST_PTR(ret_type, get_arg, get_type)    \
+    template <>                                       \
+    std::shared_ptr<ret_type> RPCHelpMan::get_arg(size_t i) const \
+    {                                                 \
+        auto result {ArgOrDefault(m_args, m_req, i)}; \
+        if (!result) return nullptr;                  \
+        auto& ref = result->get_type();               \
+        if constexpr (std::is_lvalue_reference_v<decltype(ref)>) { \
+            return std::shared_ptr<ret_type>(&ref, [](ret_type*) {}); \
+        } else { \
+            return std::make_shared<ret_type>(std::move(ref)); \
+        } \
+    }                                                 \
     void force_semicolon()
 
+
+
 TMPL_INST(int, ArgValue<int>, getInt<int>);
 TMPL_INST(uint64_t, ArgValue<uint64_t>, getInt<uint64_t>);
 TMPL_INST(const std::string&, ArgRef<std::string>, get_str);
+TMPL_INST_PTR(const std::string, ArgPtr<const std::string>, get_str);
 
 bool RPCHelpMan::IsValidNumArgs(size_t num_args) const
 {
diff --git a/src/rpc/util.h b/src/rpc/util.h
index 0ef019a780..563a619208 100644
--- a/src/rpc/util.h
+++ b/src/rpc/util.h
@@ -18,6 +18,7 @@
 #include <util/check.h>
 
 #include <string>
+#include <type_traits>
 #include <variant>
 #include <vector>
 
@@ -392,6 +393,8 @@ public:
     {
         if constexpr (std::is_integral_v<R>) {
             return ArgValue<R>(i);
+        } else if constexpr (std::is_pointer_v<R>) {
+            return ArgPtr<std::remove_pointer_t<R>>(i);
         } else {
             return ArgRef<R>(i);
         }
@@ -417,6 +420,8 @@ private:
     R ArgValue(size_t i) const;
     template <typename R>
     const R& ArgRef(size_t i) const;
+    template <typename R>
+    std::shared_ptr<R> ArgPtr(size_t i) const;
 };
 
 /**
diff --git a/src/wallet/rpc/spend.cpp b/src/wallet/rpc/spend.cpp
index b695d4bed3..0e68aff52e 100644
--- a/src/wallet/rpc/spend.cpp
+++ b/src/wallet/rpc/spend.cpp
@@ -271,8 +271,11 @@ RPCHelpMan sendtoaddress()
 
     // Wallet comments
     mapValue_t mapValue;
-    if (!request.params[2].isNull() && !request.params[2].get_str().empty())
-        mapValue["comment"] = request.params[2].get_str();
+    if (auto comment{self.Arg<const std::string*>(2)}; comment && !(comment->empty())) {
+        mapValue["comment"] = *comment;
+    }
     if (!request.params[3].isNull() && !request.params[3].get_str().empty())
         mapValue["to"] = request.params[3].get_str();
 
```
</details>

I'm not married to this approach, and it doesn't need to be added in this PR either. But I think being able to use a consistent interface for `Arg` for all types of parameters would be nice.

Reply:
> It is not possible to "move" a const reference into a shared_ptr. It will be a copy again.

Sorry, I messed up my git diff a bit, posting a slightly older version, `auto& ref = result->get_type();` has to be `auto&& ref = result->get_type();`  (updated in the above diff now). With that, I think `std::move(ref)` is only called in case of an rvalue, and we use the (from what I understand without copy) `std::shared_ptr` constructor in case of an lvalue ref. I think that avoids unnecessary copies?

> Or rather, std::optional<number> for stuff that is returned by value and const String* for stuff that is returned by reference?

That was my initial idea too, but I thought having a uniform interface (i.e. pointer for everything where it's not guaranteed to get a value) is quite a bit nicer than having to switch between std::optional and pointer depending on how we've implemented UniValue to return the value?  Although I suspect most callsites will just be using the boolean and dereference operators, which are the same for both (ignoring optional's commonly used `.value()` and `.value_or()` operators).

> Simply using a raw pointer in your approach seems fine, though.

But if `get_type()` returns by value, how would we use a raw pointer? Who owns the underlying? That's why I went for `shared_ptr`, but I'm still easily confused by memory management so apologies if I'm being dim.

Reply:
> Pretty sure it doesn't, when get_type() returns a const std::string&, both auto& and auto&& will just be a const std::string& as well.

Oh yes, I agree. So in that case, `ref` will be ` const std::string&` lvalue, and we'll use `std::shared_ptr<ret_type>(&ref, [](ret_type*) {});` where I don't think any copies happen?

> `std::optional` can be used to own optional memory

I know, I was answering your suggestion of using raw pointers instead of shared_ptr. `std::optional` solves the memory management issues, but I was just trying to avoid it to keep the interface consistent (i.e. not use optional for one thing and pointers for another) while avoiding unnecessary copies, which I think is only possible with pointers.

Reply:
That's a good point. To confirm my understanding: the lifecycles are identical for both the (1) hybrid std::optional/pointer and the (2) uniform pointer approach (i.e. in both cases, reference types can only be used during the RPC call, value types can be used whenever), but for (2) the developer (user) would have to dive into the implementation to figure out what the exact lifecycle is, whereas if we use different types it's obvious from just looking at the return type.

I'm going to think about it a bit more, but currently I'm leaning towards agreeing that the interface "inconsistency"  as per the current implementation then actually becomes a good thing. Thanks.

Reply:
> But this also may be more confusing than it helping.

Yeah, I don't really think that helps. Still no strong view, I like your current implementation, it's less complex than what I suggested, so barring strong conviction that's good enough reason for me to keep it as is. Perhaps useful to keep this convo open for future reference, but considered resolved from my end.